src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php

```diff
@@ -205,6 +251,135 @@ public function toggleStatusAction($currencyId)
         return $this->redirectToRoute('admin_currencies_index');
     }
 
+    /**
+     * Updates exchange rates.
+     *
+     * @AdminSecurity(
+     *     "is_granted('update', request.get('_legacy_controller'))",
+     *     redirectRoute="admin_currencies_index",
+     *     message="You do not have permission to edit this."
+     * )
+     * @DemoRestricted(redirectRoute="admin_currencies_index")
+     *
+     * @return RedirectResponse
+     */
+    public function updateExchangeRatesAction()
+    {
+        try {
+            $this->getCommandBus()->handle(new RefreshExchangeRatesCommand());
+
+            $this->addFlash('success', $this->trans('Successful update.', 'Admin.Notifications.Success'));
+        } catch (CannotRefreshExchangeRatesException $exception) {
+            $this->addFlash('error', $exception->getMessage());
+        }
+
+        return $this->redirectToRoute('admin_currencies_index');
+    }
+
+    /**
+     * Handles ajax request which updates live exchange rates.
+     *
+     * @param Request $request
+     *
+     * @return JsonResponse
+     */
+    public function updateLiveExchangeRatesAction(Request $request)
+    {
+        if ($this->isDemoModeEnabled()) {
+            return $this->json([
+                    'status' => false,
+                    'message' => $this->getDemoModeErrorMessage(),
+                ],
+                Response::HTTP_UNAUTHORIZED
+            );
+        }
+
+        $authLevel = $this->authorizationLevel($request->attributes->get('_legacy_controller'));
+
+        if (!in_array($authLevel, [PageVoter::LEVEL_UPDATE, PageVoter::LEVEL_DELETE])) {
+            return $this->json([
+                    'status' => false,
+                    'message' => $this->trans(
+                        'You do not have permission to update this.',
```

Comment:
And there it should be _You need permission to update this._, if we apply the #12235 recommendation.